File: src/day 3/solutions.ts
import { showResult } from "../utils";

const defaultInput = await Bun.file("input.txt").text();

const matchMultiplications = (input: string): RegExpExecArray[] => {
	const regexp = /mul\((\d+),(\d+)\)/gm;
	return [...input.matchAll(regexp)];
};

export const part1 = (input: string = defaultInput): number => {
	const res = matchMultiplications(input).reduce((acc, match) => {
		const [_full, a, b] = match;
		return acc + Number.parseInt(a) * Number.parseInt(b);
	}, 0);
	return res;
};

export const part2 = (input: string = defaultInput): number => {
	const regexp = /mul\((\d+),(\d+)\)|do\(\)|don't\(\)/gm;
	let enabled = true;
	const matchDoDont = (acc: number, match: RegExpExecArray) => {
		const [full, a, b] = match;
		if (full === "do()") {
			enabled = true;
		} else if (full === "don't()") {
			enabled = false;
		} else if (enabled) {
			return acc + Number.parseInt(a) * Number.parseInt(b);
		}
		return acc;
	};
	const res = [...input.matchAll(regexp)].reduce(matchDoDont, 0);
	return res;
};
showResult(part1(), 1);
showResult(part2(), 2);

File: src/day 3/solutions.py
import re
from pyaoc import Env

e = Env(3)
e.T("xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))", 161, None)
e.T("xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))", 161, 48)


def part1(input) -> int:
    total = 0
    regexp = re.compile(r'mul\((\d+),(\d+)\)')
    for ln in input.get_valid_lines():
        muls = regexp.findall(ln)
        for m in muls:
            total += int(m[0]) * int(m[1])
    return total


e.run_tests(1, part1)
e.run_main(1, part1)


def part2(input) -> int:
    total = 0
    regexp = re.compile(r"mul\((\d+),(\d+)\)|(do\(\))|(don't\(\))")
    enabled = True
    for ln in input.get_valid_lines():
        matches = regexp.findall(ln)
        for m in matches:
            if len(m[2]):
                enabled = True
            elif len(m[3]):
                enabled = False
            elif enabled:
                total += int(m[0]) * int(m[1])
    return total


e.run_tests(2, part2)
e.run_main(2, part2)